cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -284,31 +290,48 @@ impl cst::Policy {
         Option<ActionConstraint>,
         Option<ResourceConstraint>,
     ) {
+        // Tracks where the last variable in the scope ended. We'll point to
+        // this position to indicate where to fill in vars if we're missing one.
+        let mut end_of_last_var = self.effect.info.range_end();
+
         let mut vars = self.variables.iter().peekable();
         let principal = if let Some(head1) = vars.next() {
+            end_of_last_var = head1.info.range_end();
             head1.to_principal_constraint(errs)
         } else {
-            errs.push(ToASTError::MissingScopeConstraint(ast::Var::Principal));
+            errs.push(ToASTError::new(
+                ToASTErrorKind::MissingScopeConstraint(ast::Var::Principal),
+                SourceInfo::from_offset(end_of_last_var),
+            ));
             None
         };
         let action = if let Some(head2) = vars.next() {
+            end_of_last_var = head2.info.range_end();
             head2.to_action_constraint(errs)
         } else {
-            errs.push(ToASTError::MissingScopeConstraint(ast::Var::Action));
+            errs.push(ToASTError::new(
+                ToASTErrorKind::MissingScopeConstraint(ast::Var::Action),
+                SourceInfo::from_offset(end_of_last_var),
+            ));
             None
         };
         let resource = if let Some(head3) = vars.next() {
             head3.to_resource_constraint(errs)
         } else {
```

Comment:
missing an update to `end_of_last_var` here?

Reply:
It's not used later, so the updated is skipped